research/picai/prepare_data.py

```diff
@@ -0,0 +1,209 @@
+        adc_scan_path_filename = scan_filenames[[i for i, sf in enumerate(scan_filenames) if "adc" in sf][0]]
+        hbv_scan_path_filename = scan_filenames[[i for i, sf in enumerate(scan_filenames) if "hbv" in sf][0]]
+        ordered_scan_filenames = [t2w_scan_filename, adc_scan_path_filename, hbv_scan_path_filename]
+        scan_paths = [Path(os.path.join(scans_path, f)) for f in ordered_scan_filenames]
+
+        # Create sample and add to list of cases to be processed.
+        sample = PicaiCase(scan_paths, annotation_path, settings)
+        samples.append(sample)
+
+    # Preprocess list of samples, generate dataset overviews and write them to disk
+    paths_for_each_sample = preprocess(samples, DEFAULT_TRANSFORMS, num_threads=num_threads)
+    generate_dataset_json(paths_for_each_sample, overview_write_dir, splits_path)
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--scans_read_dir", type=str, required=True, help="Directory to read scans from.")
+    parser.add_argument("--annotation_read_dir", type=str, required=True, help="Directory to read annotation from.")
+    parser.add_argument("--scans_write_dir", type=str, required=True, help="Directory to write scans to.")
+    parser.add_argument("--annotation_write_dir", type=str, required=True, help="Directory to write annotation to.")
+    parser.add_argument(
+        "--overview_write_dir", type=str, required=True, help="Directory to write dataset overviews to."
+    )
+    parser.add_argument(
+        "--size",
+        type=int,
+        nargs="+",
+        required=False,
+        help="Size to convert images and annotations to. Default is to keep as is.",
+    )
+    parser.add_argument(
+        "--spacing",
+        type=float,
+        nargs="+",
+        required=False,
+        help="Spacing to convert images and annotations to. Default is to keep as is.",
+    )
+    parser.add_argument(
+        "--physical_size",
+        type=float,
+        nargs="+",
+        required=False,
+        help="Size to convert images and annotations to. Default is to keep as is.",
+    )
+    parser.add_argument(
+        "--scan_extension", type=str, required=False, default="mha", help="Directory to write scans to."
+    )
+    parser.add_argument(
+        "--annotation_extension", type=str, required=False, default="nii.gz", help="Directory to write annotation to."
+    )
+    parser.add_argument("--num_threads", type=int, default=4, help="Number of threads to use during preprocessing.")
+    parser.add_argument(
+        "--splits_path", type=str, default=None, help="The path to the json file containing the splits."
+    )
+
+    args = parser.parse_args()
+    preprare_data(
+        args.scans_read_dir,
+        args.annotation_read_dir,
+        args.scans_write_dir,
+        args.annotation_write_dir,
+        args.overview_write_dir,
+        tuple(args.size) if args.size is not None else None,
```

Comment:
I think these tuples will have a particular valid length. Perhaps worth checking them and throwing an error if they are not properly formatted?